src/vqvae/vqvae.py

```diff
@@ -0,0 +1,188 @@
+                                 out_channels=num_hiddens,
+                                 kernel_size=4,
+                                 stride=2, padding=1)
+        self._conv_3 = nn.Conv2d(in_channels=num_hiddens,
+                                 out_channels=num_hiddens,
+                                 kernel_size=3,
+                                 stride=1, padding=1)
+        self._residual_stack = ResidualStack(in_channels=num_hiddens,
+                                             num_hiddens=num_hiddens,
+                                             num_residual_layers=num_residual_layers,
+                                             num_residual_hiddens=num_residual_hiddens)
+
+    def forward(self, inputs):
+        x = self._conv_1(inputs)
+        x = F.relu(x)
+        
+        x = self._conv_2(x)
+        x = F.relu(x)
+        
+        x = self._conv_3(x)
+        return self._residual_stack(x)
+
+
+class Decoder(nn.Module):
+    def __init__(self, in_channels, num_hiddens, num_residual_layers, num_residual_hiddens):
+        super(Decoder, self).__init__()
+        
+        self._conv_1 = nn.Conv2d(in_channels=in_channels,
+                                 out_channels=num_hiddens,
+                                 kernel_size=3, 
+                                 stride=1, padding=1)
+        
+        self._residual_stack = ResidualStack(in_channels=num_hiddens,
+                                             num_hiddens=num_hiddens,
+                                             num_residual_layers=num_residual_layers,
+                                             num_residual_hiddens=num_residual_hiddens)
+        
+        self._conv_trans_1 = nn.ConvTranspose2d(in_channels=num_hiddens, 
+                                                out_channels=num_hiddens//2,
+                                                kernel_size=4, 
+                                                stride=2, padding=1)
+        
+        self._conv_trans_2 = nn.ConvTranspose2d(in_channels=num_hiddens//2, 
+                                                out_channels=3,
+                                                kernel_size=4, 
+                                                stride=2, padding=1)
+
+    def forward(self, inputs):
+        x = self._conv_1(inputs)
+        
+        x = self._residual_stack(x)
+        
+        x = self._conv_trans_1(x)
+        x = F.relu(x)
+        
+        return self._conv_trans_2(x)
+
+
+
+class VQVAE(pl.LightningModule):
+    
+    def __init__(self):
+        super().__init__()
+        self.encoder = Encoder(
+            in_channels=3,
+            num_hiddens=128,
+            num_residual_layers=2,
+            num_residual_hiddens=32,
+        )
+        self.conv_emb = nn.Conv2d(
+            in_channels=128,
+            out_channels=128,
+            kernel_size=1,
+            stride=1)
+        self.embeddings = nn.Embedding(512, 128)
```

Comment:
From the paper: `In our experiments we define N discrete latents (e.g., we use a field of 32 x 32 latents for ImageNet,
or 8 x 8 x 10 for CIFAR10)`. Do they mean then that their discrete latents have a dimensionality of `640`?

Reply:
Because for ImageNet, they later say that their discrete space is `32 x 32 x 1` with `K=512` i.e., their `D = 32 x 32 x 1` = 1024, right?

Reply:
So 8x8x10 actually means they get 10 different codes for each patch, using 10 different codebooks. I haven't implemented multiple codebooks yet.